src/dense.jl

```diff
@@ -39,15 +39,18 @@ struct AutoDiffractor <: AbstractADType end
 mode(::AutoDiffractor) = ForwardOrReverseMode()
 
 """
-    AutoEnzyme{M}
+    AutoEnzyme{M,constant_function}
 
 Struct used to select the [Enzyme.jl](https://github.com/EnzymeAD/Enzyme.jl) backend for automatic differentiation.
 
 Defined by [ADTypes.jl](https://github.com/SciML/ADTypes.jl).
 
 # Constructors
 
-    AutoEnzyme(; mode=nothing)
+    AutoEnzyme(; mode=nothing, constant_function::Bool=false)
+
+The `constant_function` keyword argument (and type parameter) determines whether the function object itself should be considered constant or not during differentiation with Enzyme.jl.
+For simple functions, `constant_function` should usually be set to `false`, but in the case of closures or callable structs which contain differentiated data that can be treated as constant, `constant_function` should be set to `true` for increased performance.
```

Comment:
```suggestion
For simple functions, `constant_function` should usually be set to `false`, but in the case of closures or callable structs which contain differentiated data that can be treated as constant, `constant_function` should be set to `true` for increased performance.

!!! note
    If `constant_function` is set to true but the enclosed data is not truly constant, then Enzyme will not compute the
    correct derivative values. An example of such a function is:
    
    ```julia
    cache = [0.0]
    function f(x)
       cache[1] = x[1]^2
       cache[1] + x[1]
    end
    ```

    In this case, the enclosed cache is a function of the differentiated input, and thus its values are non-constant with
    respect to the input. Thus in order to compute the correct derivative of the output, the derivative must propagate
    through the `cache` value and it must not be treated as constant. Alternatively, the following function can treat
    `parameter` as a constant:

    ```julia
    parameter = [0.0]
    function f(x)
       parameter[1] + x[1]
    end
    ```

    because `parameter` is never modified based on the input `x`. In this case, `constant_function = true` would allow
    the chosen differentiation system to perform extra memory and compute optimizations under the assumption
    `parameter` is kept constant. 

```